numpyro/infer/einstein/stein.py

```diff
@@ -0,0 +1,342 @@
+                             for random (EXPERIMENTAL)
+        :param sp_mode: Stein Point MCMC mode for calculating Kernelized Stein Discrepancy. Either 'local'
+                        for only the updated MCMC particles or 'global' for all particles. (EXPERIMENTAL)
+        :param num_mcmc_particles: Number of particles that should be updated with Stein Point MCMC
+                                   (should be a subset of number of Stein particles) (EXPERIMENTAL)
+        :param num_mcmc_warmup: Number of warmup steps for the MCMC sampler (EXPERIMENTAL)
+        :param num_mcmc_updates: Number of MCMC update steps at each iteration (EXPERIMENTAL)
+        :param sampler_fn: The MCMC sampling kernel used for the Stein Point MCMC updates (EXPERIMENTAL)
+        :param sampler_kwargs: Keyword arguments provided to the MCMC sampling kernel (EXPERIMENTAL)
+        :param mcmc_kwargs: Keyword arguments provided to the MCMC interface (EXPERIMENTAL)
+        :param static_kwargs: Static keyword arguments for the model / guide, i.e. arguments
+            that remain constant during fitting.
+        """
+        super().__init__(model, guide, optim, loss, name='Stein', **static_kwargs)
+        assert sp_mcmc_crit == 'infl' or sp_mcmc_crit == 'rand'
+        assert sp_mode == 'local' or sp_mode == 'global'
+        assert 0 <= num_mcmc_particles <= num_particles
+
+        self._inference_model = model
+        self.model = model
+        self.guide = guide
+        self.optim = optim
+        self.loss = loss
+        self.kernel_fn = kernel_fn
+        self.static_kwargs = static_kwargs
+        self.num_particles = num_particles
+        self.loss_temperature = loss_temperature
+        self.repulsion_temperature = repulsion_temperature
+        self.enum = enum
+        self.classic_guide_params_fn = classic_guide_params_fn
+        self.sp_mcmc_crit = sp_mcmc_crit
+        self.sp_mode = sp_mode
+        self.num_mcmc_particles = num_mcmc_particles
+        self.num_mcmc_warmup = num_mcmc_warmup
+        self.num_mcmc_updates = num_mcmc_updates
+        self.sampler_fn = sampler_fn
+        self.sampler_kwargs = sampler_kwargs or dict()
+        self.mcmc_kwargs = mcmc_kwargs or dict()
+        self.mcmc: MCMC = None
+        self.guide_param_names = None
+        self.constrain_fn = None
+        self.uconstrain_fn = None
+        self.particle_transform_fn = None
+
+    def _apply_kernel(self, kernel, x, y, v):
+        if self.kernel_fn.mode == 'norm' or self.kernel_fn.mode == 'vector':
+            return kernel(x, y) * v
+        else:
+            return kernel(x, y) @ v
+
+    def _kernel_grad(self, kernel, x, y):
+        if self.kernel_fn.mode == 'norm':
+            return jax.grad(lambda x: kernel(x, y))(x)
+        elif self.kernel_fn.mode == 'vector':
+            return jax.vmap(lambda i: jax.grad(lambda x: kernel(x, y)[i])(x)[i])(jnp.arange(x.shape[0]))
```

Comment:
What is the shape of `x`, `y` and `kernel(x, y)` here? It seems to be that we can computing `kernel(x, y)` multiple times, across different `i`, which can be avoided by using `jacfwd`, `jacfwd`, or `vmap` directly over `x` I guess (I still need to understand the shapes first).

Reply:
The kernel_fn to compute kernel for pair of particles.
                 Modes: norm `(d,) (d,)-> ()`,  vector `(d,) (d,) -> (d)`, or matrix `(d,) (d,) -> (d,d)`

Reply:
From SteinKernel 😄